fscryptctl.c

```diff
@@ -166,355 +161,557 @@ const char *policy_error(int errno_val) {
+    ret = 0;
+    goto cleanup;
+  }
+  memcpy(raw_key, buf, ret);
+cleanup:
+  secure_wipe(buf, sizeof(buf));
   return ret;
 }
 
-static int get_policy(const char *path, struct fscrypt_policy *policy) {
-  // We can query the policy for a directory or a file in that directory.
-  int fd = open(path, O_RDONLY);
+static bool get_policy(const char *path,
+                       struct fscrypt_get_policy_ex_arg *arg) {
+  int fd = open(path, O_RDONLY | O_CLOEXEC);
   if (fd < 0) {
-    return -1;
+    fprintf(stderr, "error: opening %s: %s\n", path, strerror(errno));
+    return false;
   }
 
-  int ret = ioctl(fd, FS_IOC_GET_ENCRYPTION_POLICY, policy);
+  arg->policy_size = sizeof(arg->policy);
+  int ret = ioctl(fd, FS_IOC_GET_ENCRYPTION_POLICY_EX, arg);
   close(fd);
 
-  if (ret < 0) {
-    // Kernels prior to v4.11 returned ENOENT if the file did not have an
-    // encryption policy, newer kernels properly return ENODATA. This lets us
-    // print the right error in policy_error regardless of kernel version.
-    if (errno == ENOENT) {
-      errno = ENODATA;
-    }
+  if (ret != 0) {
+    fprintf(stderr, "error: getting policy for %s: %s\n", path,
+            describe_get_policy_error(errno));
+    return false;
   }
-
-  return ret;
+  return true;
 }
 
-static int set_policy(const char *path, const struct fscrypt_policy *policy) {
-  // Policies can only be set on directories
-  int fd = open(path, O_RDONLY | O_DIRECTORY);
+static bool set_policy(const char *path,
+                       const struct fscrypt_policy_v2 *policy) {
+  int fd = open(path, O_RDONLY | O_CLOEXEC);
   if (fd < 0) {
-    return -1;
+    fprintf(stderr, "error: opening %s: %s\n", path, strerror(errno));
+    return false;
   }
 
   int ret = ioctl(fd, FS_IOC_SET_ENCRYPTION_POLICY, policy);
   close(fd);
 
-  return ret;
+  if (ret != 0) {
+    fprintf(stderr, "error: setting policy for %s: %s\n", path,
+            describe_set_policy_error(errno));
+    return false;
+  }
+  return true;
 }
 
-/* Functions for various actions, return 0 on success, non-zero on failure. */
+// -----------------------------------------------------------------------------
+//                                 Commands
+// -----------------------------------------------------------------------------
 
-// Get the descriptor for some key data passed via stdin. Provided key data must
-// have length FS_MAX_KEY_SIZE. Output will be formatted as hex.
-static int cmd_get_descriptor(int argc, char *const argv[]) {
-  if (argc != 2) {
-    fputs("error: unexpected arguments\n", stderr);
+static int cmd_add_key(int argc, char *const argv[]) {
+  handle_no_options(&argc, &argv);
+  if (argc != 1) {
+    fputs("error: must specify a single mountpoint\n", stderr);
     return EXIT_FAILURE;
   }
+  const char *mountpoint = argv[0];
 
-  int ret = EXIT_SUCCESS;
-  uint8_t key[FS_MAX_KEY_SIZE];
+  struct fscrypt_add_key_arg *arg =
+      calloc(sizeof(*arg) + FSCRYPT_MAX_KEY_SIZE, 1);
```

Comment:
Is there a reason we use calloc here instead of just a stack allocated buffer?

Reply:
We need to ensure that the buffer is properly aligned for the data structure, and calloc() is the easiest way to do that.

Reply:
Makes sense